onmt/Models.py

```diff
@@ -113,55 +134,72 @@ def forward(self, src_input):
                                  .expand_as(emb))
             emb = self.dropout(emb)
 
-        out_length, out_batch, emb_size = emb.size()
+        out_length, out_batch, emb_dim = emb.size()
         aeq(in_length, out_length)
         aeq(in_length, out_length)
-        aeq(emb_size, self.embedding_size)
+        aeq(emb_dim, self.embedding_dim)
 
         return emb
 
+    @staticmethod
+    def create_instance(opt, dicts, for_encoder, feature_dicts=None):
+        """
+        Create an Embeddings instance.
+        """
+        if for_encoder:
+            embedding_dim = opt.src_word_vec_size
+        else:
+            embedding_dim = opt.tgt_word_vec_size
+        return Embeddings(embedding_dim,
+                          opt.position_encoding,
+                          opt.feat_merge,
+                          opt.feat_vec_exponent,
+                          opt.feat_vec_size,
+                          opt.dropout,
+                          dicts, feature_dicts)
+
 
 class Encoder(nn.Module):
     """
     Encoder recurrent neural network.
     """
-    def __init__(self, opt, dicts, feature_dicts=None):
+    def __init__(self, encoder_type, brnn, rnn_type,
+                 num_layers, rnn_size, dropout, enc_emb):
         """
         Args:
-            opt: Model options.
-            dicts (`Dict`): The src dictionary
-            features_dicts (`[Dict]`): List of src feature dictionaries.
+            encoder_type (string): rnn, brnn, mean, or transformer.
+            brnn (bool): brnn Encoder. Prefer `encoder_type=brnn` now.
+            rnn_type (string): LSTM or GRU.
+            num_layers (int): number of Encoder layers.
+            rnn_size (int): size of hidden states of a rnn.
+            dropout (float): dropout probablity.
+            enc_emb (Embeddings): vocab embeddings for this Encoder.
         """
-        # Number of rnn layers.
-        self.layers = opt.enc_layers
-
-        # Use a bidirectional model.
-        self.num_directions = 2 if opt.brnn else 1
-        assert opt.rnn_size % self.num_directions == 0
-
-        # Size of the encoder RNN.
-        self.hidden_size = opt.rnn_size // self.num_directions
-
+        # Call nn.Module.__init().
         super(Encoder, self).__init__()
```

Comment:
Actually I think it is important that non-modules go before the constructor? Otherwise it saves them and breaks loading.